cli/flox-watchdog/src/process.rs

```diff
@@ -0,0 +1,414 @@
+    //       share anything behind #[cfg(test)] across crates
+
+    /// Start a shortlived process that we can check the PID is running.
+    pub fn start_process() -> Child {
+        Command::new("sleep")
+            .arg("2")
+            .spawn()
+            .expect("failed to start")
+    }
+
+    /// Stop a shortlived process that we can check the PID is not running. It's
+    /// unlikely, but not impossible, that the kernel will have not re-used the
+    /// PID by the time we check it.
+    pub fn stop_process(mut child: Child) {
+        child.kill().expect("failed to kill");
+        child.wait().expect("failed to wait");
+    }
+
+    /// Makes two Arc<AtomicBool>s to mimic the shutdown flags used by
+    /// the watchdog
+    pub fn shutdown_flags() -> (Arc<AtomicBool>, Arc<AtomicBool>) {
+        (
+            Arc::new(AtomicBool::new(false)),
+            Arc::new(AtomicBool::new(false)),
+        )
+    }
+
+    /// Writes a registry to a temporary file, adding an entry for the provided
+    /// path hash
+    pub fn path_for_registry_with_entry(path_hash: impl AsRef<str>) -> NamedTempFile {
+        let path = NamedTempFile::new().unwrap();
+        let mut reg = EnvRegistry::default();
+        let entry = RegistryEntry {
+            path_hash: String::from(path_hash.as_ref()),
+            path: PathBuf::from("foo"),
+            envs: vec![],
+            activations: HashSet::new(),
+        };
+        reg.entries.push(entry);
+        let string = serde_json::to_string(&reg).unwrap();
+        std::fs::write(&path, string).unwrap();
+        path
+    }
+
+    /// Wait some attempts for the process to reach the desired state
+    fn poll_until_state(state: ProcStatus, pid: ActivationPid) {
+        for _ in 0..10 {
+            if PidWatcher::read_pid_status(pid) == state {
+                return;
+            }
+            std::thread::sleep(Duration::from_millis(10));
+        }
+        panic!("never entered zombie state");
+    }
+
+    #[test]
+    fn reports_that_pid1_is_running() {
+        assert!(PidWatcher::pid_is_running(1.into()));
+    }
+
+    #[test]
+    fn detects_running_or_not_running_process() {
+        let proc = start_process();
+        let pid = proc.id() as i32;
+        assert!(PidWatcher::pid_is_running(pid.into()));
+        stop_process(proc);
+        assert!(!PidWatcher::pid_is_running(pid.into()));
+    }
+
+    #[test]
+    fn detects_zombie() {
+        let mut proc = Command::new("true").spawn().unwrap();
+        let pid = proc.id() as i32;
+        poll_until_state(ProcStatus::Zombie, pid.into());
+        assert!(!PidWatcher::pid_is_running(pid.into()));
+        assert_eq!(PidWatcher::read_pid_status(pid.into()), ProcStatus::Zombie);
+        proc.wait().unwrap();
+    }
+
+    #[test]
+    fn terminates_when_all_pids_terminate() {
+        let proc1 = start_process();
+        let pid1 = ActivationPid::from(proc1.id() as i32);
+        let proc2 = start_process();
```

Comment:
question nonblocking: what's proc2 used for in this test?

Reply:
The watcher is initialized with one PID. The second process is there to ensure that the watcher doesn't treat it as special i.e. it makes sure that the watcher waits for _all_ processes not just the one it's initialized with.